sql: fix interactions of PARTITION BY NOTHING and PARTITION BY ALL

otan opened the pull request:
Before this PR, specifying PARTITION BY NOTHING whilst having PARTITION
BY ALL would be allowed. However, upon closer inspection, we shouldn't
allow someone to specify PARTITION BY NOTHING if PARTITION ALL BY is
defined on the table. This PR introduces proper checks for this.

Furthermore, introduce support for PARTITION ALL BY NOTHING, which
ensures the entire table has no PARTITION BY.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59168)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59168)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_all_by_nothing, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59168#-MRWum815jC3uMHAt6vY:-MRWum815jC3uMHAt6vZ:b-u65sl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/034db9e23fb86bdbad4920602d5aab92c1f83d37/pkg/ccl/logictestccl/testdata/logic_test/partitioning_all_by_nothing#L6)):*
> ```
> SET experimental_enable_implicit_column_partitioning = true
> 
> statement error cannot define PARTITION BY on an unique constraint if the table has a PARTITION ALL BY definition
> ```

Super Nit: Is this case tested elsewhere?  Vague memories...
___
*[pkg/sql/create_table.go, line 1387 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59168#-MRZzexVDRkTMJwxWONB:-MRZzexVDRkTMJwxWONC:b7nmuqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/034db9e23fb86bdbad4920602d5aab92c1f83d37/pkg/sql/create_table.go#L1387)):*
> ```Go
> 		}
> 
> 		// Check no PARTITION BY is set on any field.
> ```

Super Nit: any field?  or any column?  Since this is the PartitionByTable, I'd think that it would only relate to table level partitioning.  Or maybe we can just say "Check no PARTITION BY is set."?
___
*[pkg/sql/create_table.go, line 1388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59168#-MR_-Hyl0iyYGLdCz9OT:-MR_-Hyl0iyYGLdCz9OU:b-osrlcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/034db9e23fb86bdbad4920602d5aab92c1f83d37/pkg/sql/create_table.go#L1388)):*
> ```Go
> 
> 		// Check no PARTITION BY is set on any field.
> 		if n.PartitionByTable != nil {
> ```

Nit: Could we create ContainsPartitioningClause() to differentiate between containing partitions and PARTITION BY NOTHING?  The meaning of `!= nil` is not immediately obvious when reading this function.


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix interactions of PARTITION BY NOTHING and PARTITION BY ALL

Before this PR, specifying PARTITION BY NOTHING whilst having PARTITION
BY ALL would be allowed. However, upon closer inspection, we shouldn't
allow someone to specify PARTITION BY NOTHING if PARTITION ALL BY is
defined on the table. This PR introduces proper checks for this.

Furthermore, introduce support for PARTITION ALL BY NOTHING, which
ensures the entire table has no PARTITION BY.

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59168)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_all_by_nothing, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59168#-MRWum815jC3uMHAt6vY:-MRaUWa7DfkwhkvoBNQQ:b-yoeaxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/034db9e23fb86bdbad4920602d5aab92c1f83d37/pkg/ccl/logictestccl/testdata/logic_test/partitioning_all_by_nothing#L6)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Super Nit: Is this case tested elsewhere?  Vague memories...
</blockquote></details>

not PARTITION BY NOTHING
___
*[pkg/sql/create_table.go, line 1387 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59168#-MRZzexVDRkTMJwxWONB:-MRaUZAA3gFotKkYcDvd:b-2ppqol) ([raw file](https://github.com/cockroachdb/cockroach/blob/034db9e23fb86bdbad4920602d5aab92c1f83d37/pkg/sql/create_table.go#L1387)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Super Nit: any field?  or any column?  Since this is the PartitionByTable, I'd think that it would only relate to table level partitioning.  Or maybe we can just say "Check no PARTITION BY is set."?
</blockquote></details>

technically column definition. done!


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm

otan commented:
bors r=ajstorm
